libcaf_core/caf/uri.cpp

```diff
@@ -250,7 +250,14 @@ std::string to_string(const uri& x) {
 std::string to_string(const uri::authority_type& x) {
   std::string str;
   if (!x.userinfo.empty()) {
-    uri::encode(str, x.userinfo);
+    std::string_view userinfo = x.userinfo;
+    // TODO: with this approach, username can't contain a ':'.
+    auto pos = userinfo.find(':');
+    uri::encode(str, userinfo.substr(0, pos));
+    if (pos != std::string::npos) {
+      str += ':';
+      uri::encode(str, userinfo.substr(pos + 1));
+    }
```

Comment:
Storing user and password in a single string makes it impossible to disambiguate what a `:` means. We need to change the `uri` class for this:

```c++
  /// Host subcomponent of the authority component. Either an IP address or
  /// an hostname as string.
  using host_type = std::variant<std::string, ip_address>;

  /// Userinfo subcomponent of the authority component. If present, contains the user name and optionally a password.
  struct userinfo_type {
    std::string name;
    std::optional<std::string> password;
  };

  /// Bundles the authority component of the URI, i.e., userinfo, host, and
  /// port.
  struct CAF_CORE_EXPORT authority_type {
    std::optional<userinfo_type> userinfo;
    host_type host;
    uint16_t port;
```

It's a bit awkward to have the double `optional`, but otherwise we can't distinguish between `foo://localhost` (no userinfo), `foo://@localhost` (empty username but no password), and `foo://:@localhost` (empty username and empty password).

Reply:
I assumed we needed to split this, I've just wanted to get your feedback on how much refactoring is permissible here. What you're suggesting is also aligned with the [URI RFC](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.1):

> The userinfo subcomponent may consist of a user name and, optionally, scheme-specific information about how to gain authorization to access the resource.